File: Project-source/forge-verifier/java/src/main/java/com/yourco/forge/verifier/CrlClient.java
package com.yourco.forge.verifier;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Instant;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * CRL client — fetches and caches {@code GET /api/v1/revocation-list}.
 *
 * <p>Honors {@code ETag} / {@code If-None-Match}: on 304 we keep the existing
 * snapshot. Thread-safe; reuse a single instance across verify calls.
 */
public final class CrlClient {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final String baseUrl;
    private final String apiKey;
    private final String userAgent;
    private final HttpClient http;

    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
    private Set<String> revoked = new HashSet<>();
    private String etag = "";
    private Instant lastFetched = Instant.EPOCH;

    public CrlClient(String baseUrl, String apiKey, String userAgent) {
        this(baseUrl, apiKey, userAgent, HttpClient.newHttpClient());
    }

    public CrlClient(String baseUrl, String apiKey, String userAgent, HttpClient http) {
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
        this.userAgent = userAgent;
        this.http = http;
    }

    public void refresh() throws IOException, InterruptedException {
        HttpRequest.Builder req = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/revocation-list"))
                .header("Accept", "application/json")
                .GET();
        if (apiKey != null && !apiKey.isEmpty()) req.header("X-Forge-API-Key", apiKey);
        if (userAgent != null && !userAgent.isEmpty()) req.header("User-Agent", userAgent);
        lock.readLock().lock();
        try {
            if (!etag.isEmpty()) req.header("If-None-Match", etag);
        } finally {
            lock.readLock().unlock();
        }

        HttpResponse<byte[]> resp = http.send(req.build(), HttpResponse.BodyHandlers.ofByteArray());
        int code = resp.statusCode();
        if (code == 304) {
            touchLastFetched();
            return;
        }
        if (code == 401) throw new IOException("crl: invalid api key (401)");
        if (code == 429) throw new IOException("crl: rate limited (429)");
        if (code >= 500) throw new IOException("crl: server error " + code);
        if (code != 200) throw new IOException("crl: unexpected status " + code);

        Body body = MAPPER.readValue(resp.body(), Body.class);
        Set<String> next = new HashSet<>(body.licenseIds == null ? java.util.List.of() : body.licenseIds);
        String newEtag = resp.headers().firstValue("ETag").orElse("");
        lock.writeLock().lock();
        try {
            revoked = next;
            etag = newEtag;
            lastFetched = Instant.now();
        } finally {
            lock.writeLock().unlock();
        }
    }

    public boolean isRevoked(String licenseId) {
        lock.readLock().lock();
        try {
            return revoked.contains(licenseId);
        } finally {
            lock.readLock().unlock();
        }
    }

    public Instant lastFetched() {
        lock.readLock().lock();
        try {
            return lastFetched;
        } finally {
            lock.readLock().unlock();
        }
    }

    public int size() {
        lock.readLock().lock();
        try {
            return revoked.size();
        } finally {
            lock.readLock().unlock();
        }
    }

    private void touchLastFetched() {
        lock.writeLock().lock();
        try {
            lastFetched = Instant.now();
        } finally {
            lock.writeLock().unlock();
        }
    }

    static final class Body {
        public java.util.List<String> licenseIds;
        public String generatedAt;
    }
}

File: Project-source/forge-verifier/csharp/src/CrlClient.cs
// CRL client — GET /api/v1/revocation-list with ETag/If-None-Match caching.
// Thread-safe via a ReaderWriterLockSlim; reuse a single instance.

using System.Text.Json;
using System.Text.Json.Serialization;

namespace YourCo.Forge.Verifier;

public sealed class CrlClient
{
    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
    };

    private readonly string _baseUrl;
    private readonly string _apiKey;
    private readonly string _userAgent;
    private readonly HttpClient _http;

    private readonly ReaderWriterLockSlim _lock = new();
    private HashSet<string> _revoked = new();
    private string _etag = "";
    private DateTimeOffset _lastFetched;

    public CrlClient(string baseUrl, string apiKey, string userAgent, HttpClient? http = null)
    {
        _baseUrl = baseUrl;
        _apiKey = apiKey;
        _userAgent = userAgent;
        _http = http ?? new HttpClient();
    }

    public async Task RefreshAsync(CancellationToken ct = default)
    {
        using var req = new HttpRequestMessage(HttpMethod.Get, $"{_baseUrl}/revocation-list");
        req.Headers.TryAddWithoutValidation("Accept", "application/json");
        if (!string.IsNullOrEmpty(_apiKey)) req.Headers.TryAddWithoutValidation("X-Forge-API-Key", _apiKey);
        if (!string.IsNullOrEmpty(_userAgent)) req.Headers.TryAddWithoutValidation("User-Agent", _userAgent);
        _lock.EnterReadLock();
        try { if (!string.IsNullOrEmpty(_etag)) req.Headers.TryAddWithoutValidation("If-None-Match", _etag); }
        finally { _lock.ExitReadLock(); }

        using var resp = await _http.SendAsync(req, ct).ConfigureAwait(false);
        var code = (int)resp.StatusCode;
        if (code == 304)
        {
            TouchLastFetched();
            return;
        }
        if (code == 401) throw new InvalidOperationException("crl: invalid api key (401)");
        if (code == 429) throw new InvalidOperationException("crl: rate limited (429)");
        if (code >= 500) throw new InvalidOperationException($"crl: server error {code}");
        if (code != 200) throw new InvalidOperationException($"crl: unexpected status {code}");

        var stream = await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
        var body = (await JsonSerializer.DeserializeAsync<CrlBody>(stream, JsonOpts, ct)
            .ConfigureAwait(false))!;
        var next = new HashSet<string>(body.LicenseIds ?? new List<string>());
        var newEtag = resp.Headers.ETag?.Tag ?? "";

        _lock.EnterWriteLock();
        try
        {
            _revoked = next;
            _etag = newEtag;
            _lastFetched = DateTimeOffset.UtcNow;
        }
        finally { _lock.ExitWriteLock(); }
    }

    public bool IsRevoked(string licenseId)
    {
        _lock.EnterReadLock();
        try { return _revoked.Contains(licenseId); }
        finally { _lock.ExitReadLock(); }
    }

    public DateTimeOffset LastFetched
    {
        get
        {
            _lock.EnterReadLock();
            try { return _lastFetched; }
            finally { _lock.ExitReadLock(); }
        }
    }

    public int Size
    {
        get
        {
            _lock.EnterReadLock();
            try { return _revoked.Count; }
            finally { _lock.ExitReadLock(); }
        }
    }

    private void TouchLastFetched()
    {
        _lock.EnterWriteLock();
        try { _lastFetched = DateTimeOffset.UtcNow; }
        finally { _lock.ExitWriteLock(); }
    }

    private sealed class CrlBody
    {
        [JsonPropertyName("license_ids")] public List<string>? LicenseIds { get; set; }
        [JsonPropertyName("generated_at")] public string? GeneratedAt { get; set; }
    }
}
